File: packages/pulse/js/src/messages.ts
// =================================================================
// Message Types
// =================================================================

import type { RouteInfo } from "./helpers";
import type { VDOM, VDOMNode, VDOMUpdate } from "./vdom";

// Based on pulse/messages.py
export interface ServerInitMessage {
	type: "vdom_init";
	path: string;
	vdom: VDOM;
}

export interface ServerUpdateMessage {
	type: "vdom_update";
	path: string;
	ops: VDOMUpdate[];
}

export interface ServerError {
	message: string;
	stack?: string;
	phase: "render" | "callback" | "mount" | "unmount" | "navigate" | "server";
	details?: Record<string, any>;
}

export interface ServerErrorMessage {
	type: "server_error";
	path: string;
	error: ServerError;
}

export interface ServerApiCallMessage {
	type: "api_call";
	id: string;
	url: string; // absolute or relative
	method: string;
	headers: Record<string, string>;
	body: any | null;
	credentials: "include" | "omit";
}

export interface ChannelMessageBase {
	type: "channel";
	channel: string;
}

export interface ChannelConnectMessage extends ChannelMessageBase {
	action: "connect";
	subscriptionId: string;
	owner?: string;
}

export interface ChannelDisconnectMessage extends ChannelMessageBase {
	action: "disconnect";
	subscriptionId: string;
	owner?: string;
}

export interface ChannelConnectAckMessage extends ChannelMessageBase {
	action: "connect_ack";
	subscriptionId: string;
	accepted: boolean;
	error?: string;
}

export interface ChannelCloseMessage extends ChannelMessageBase {
	action: "close";
	subscriptionId: string;
	reason?: string;
}

export interface ChannelEventMessage extends ChannelMessageBase {
	action: "event";
	event: string;
	payload?: any;
	subscriptionId?: string;
}

export interface ChannelRequestMessage extends ChannelMessageBase {
	action: "request";
	event: string;
	requestId: string;
	payload?: any;
	subscriptionId?: string;
}

export interface ChannelResponseMessage extends ChannelMessageBase {
	action: "response";
	responseTo: string;
	payload?: any;
	error?: any;
	subscriptionId?: string;
}

export type ClientChannelMessage =
	| ChannelConnectMessage
	| ChannelDisconnectMessage
	| ChannelEventMessage
	| ChannelRequestMessage
	| ChannelResponseMessage;

export type ServerChannelMessage =
	| ChannelConnectAckMessage
	| ChannelCloseMessage
	| ChannelEventMessage
	| ChannelRequestMessage
	| ChannelResponseMessage;

export interface ServerNavigateToMessage {
	type: "navigate_to";
	path: string;
	replace: boolean;
	hard: boolean;
	sourceRoutePath?: string;
	sourcePath?: string;
	sourceMountId?: string;
}

export interface ServerReloadMessage {
	type: "reload";
}

export interface ServerAttachAckMessage {
	type: "attach_ack";
	path: string;
	attachId: string;
}

export interface ServerJsExecMessage {
	type: "js_exec";
	path: string;
	id: string;
	expr: VDOMNode;
}

export type ServerMessage =
	| ServerInitMessage
	| ServerUpdateMessage
	| ServerErrorMessage
	| ServerApiCallMessage
	| ServerNavigateToMessage
	| ServerReloadMessage
	| ServerAttachAckMessage
	| ServerChannelMessage
	| ServerJsExecMessage;

export interface ClientCallbackMessage {
	type: "callback";
	path: string;
	callback: string;
	args: any[];
}

export interface ClientAttachMessage {
	type: "attach";
	path: string;
	routeInfo: RouteInfo;
	attachId: string;
}
export interface ClientUpdateMessage {
	type: "update";
	path: string;
	routeInfo: RouteInfo;
}
export interface ClientDetachMessage {
	type: "detach";
	path: string;
}

export interface ClientApiResultMessage {
	type: "api_result";
	id: string;
	ok: boolean;
	status: number;
	headers: Record<string, string>;
	body: any | null;
}

export interface ClientJsResultMessage {
	type: "js_result";
	id: string;
	result: any;
	error: string | null;
}

export type ClientMessage =
	| ClientAttachMessage
	| ClientCallbackMessage
	| ClientUpdateMessage
	| ClientDetachMessage
	| ClientApiResultMessage
	| ClientChannelMessage
	| ClientJsResultMessage;

File: packages/pulse/python/src/pulse/messages.py
from typing import Any, Literal, NotRequired, TypedDict

from pulse.routing import RouteInfo
from pulse.transpiler.vdom import VDOM, VDOMNode, VDOMOperation


# ====================
# Server messages
# ====================
class ServerInitMessage(TypedDict):
	type: Literal["vdom_init"]
	path: str
	vdom: VDOM


class ServerUpdateMessage(TypedDict):
	type: Literal["vdom_update"]
	path: str
	ops: list[VDOMOperation]


ServerErrorPhase = Literal[
	"render", "callback", "mount", "unmount", "navigate", "server", "effect", "connect"
]


class ServerErrorInfo(TypedDict):
	# High-level human message
	message: str
	# Full stack trace string (server formatted)
	stack: str
	# Which phase failed
	phase: ServerErrorPhase
	# Extra details (callback key, etc.)
	details: dict[str, Any]


class ServerErrorMessage(TypedDict):
	type: Literal["server_error"]
	path: str
	error: ServerErrorInfo


class ServerNavigateToMessage(TypedDict):
	type: Literal["navigate_to"]
	path: str
	replace: bool
	hard: bool
	sourceRoutePath: NotRequired[str]
	sourcePath: NotRequired[str]
	sourceMountId: NotRequired[str]


class ServerReloadMessage(TypedDict):
	type: Literal["reload"]


class ServerAttachAckMessage(TypedDict):
	type: Literal["attach_ack"]
	path: str
	attachId: str


class ServerApiCallMessage(TypedDict):
	type: Literal["api_call"]
	# Correlation id to match request/response
	id: str
	url: str
	method: str
	headers: dict[str, str]
	# Body can be JSON-serializable or None
	body: Any | None
	# Whether to include credentials (cookies)
	credentials: Literal["include", "omit"]


class ServerChannelEventMessage(TypedDict):
	type: Literal["channel_event"]
	channel: str
	event: str
	payload: Any


class ServerChannelRequestMessage(TypedDict):
	type: Literal["channel_request"]
	channel: str
	event: str
	requestId: str
	payload: Any


class ServerChannelResponseSuccessMessage(TypedDict):
	type: Literal["channel_response"]
	channel: str
	responseTo: str
	ok: Literal[True]
	payload: Any


class ServerChannelResponseErrorMessage(TypedDict):
	type: Literal["channel_response"]
	channel: str
	responseTo: str
	ok: Literal[False]
	error: str


class ServerJsExecMessage(TypedDict):
	"""Execute JavaScript expression on the client."""

	type: Literal["js_exec"]
	path: str
	id: str
	expr: VDOMNode


# ====================
# Client messages
# ====================
class ClientCallbackMessage(TypedDict):
	type: Literal["callback"]
	path: str
	callback: str
	args: list[Any]


class ClientAttachMessage(TypedDict):
	type: Literal["attach"]
	path: str
	routeInfo: RouteInfo
	attachId: str


class ClientUpdateMessage(TypedDict):
	type: Literal["update"]
	path: str
	routeInfo: RouteInfo


class ClientDetachMessage(TypedDict):
	type: Literal["detach"]
	path: str


class ClientApiResultMessage(TypedDict):
	type: Literal["api_result"]
	id: str
	ok: bool
	status: int
	headers: dict[str, str]
	body: Any | None


class ClientChannelEventMessage(TypedDict):
	type: Literal["channel_event"]
	channel: str
	event: str
	payload: Any


class ClientChannelRequestMessage(TypedDict):
	type: Literal["channel_request"]
	channel: str
	event: str
	requestId: str
	payload: Any


class ClientChannelResponseSuccessMessage(TypedDict):
	type: Literal["channel_response"]
	channel: str
	responseTo: str
	ok: Literal[True]
	payload: Any


class ClientChannelResponseErrorMessage(TypedDict):
	type: Literal["channel_response"]
	channel: str
	responseTo: str
	ok: Literal[False]
	error: str


class ClientJsResultSuccessMessage(TypedDict):
	"""Result of client-side JS execution."""

	type: Literal["js_result"]
	id: str
	ok: Literal[True]
	result: Any


class ClientJsResultErrorMessage(TypedDict):
	"""Failure from client-side JS execution."""

	type: Literal["js_result"]
	id: str
	ok: Literal[False]
	error: str


ServerChannelResponseMessage = (
	ServerChannelResponseSuccessMessage | ServerChannelResponseErrorMessage
)
ServerChannelMessage = (
	ServerChannelEventMessage
	| ServerChannelRequestMessage
	| ServerChannelResponseMessage
)
ServerMessage = (
	ServerInitMessage
	| ServerUpdateMessage
	| ServerErrorMessage
	| ServerApiCallMessage
	| ServerNavigateToMessage
	| ServerReloadMessage
	| ServerAttachAckMessage
	| ServerChannelMessage
	| ServerJsExecMessage
)


ClientPulseMessage = (
	ClientCallbackMessage
	| ClientAttachMessage
	| ClientUpdateMessage
	| ClientDetachMessage
	| ClientApiResultMessage
	| ClientJsResultSuccessMessage
	| ClientJsResultErrorMessage
)
ClientChannelResponseMessage = (
	ClientChannelResponseSuccessMessage | ClientChannelResponseErrorMessage
)
ClientChannelIncomingMessage = ClientChannelEventMessage | ClientChannelRequestMessage
ClientChannelMessage = ClientChannelIncomingMessage | ClientChannelResponseMessage
ClientMessage = ClientPulseMessage | ClientChannelMessage


class PrerenderPayload(TypedDict):
	paths: list[str]
	routeInfo: RouteInfo
	ttlSeconds: NotRequired[float | int]
	renderId: NotRequired[str]


class SocketIODirectives(TypedDict):
	headers: dict[str, str]
	auth: dict[str, str]
	query: dict[str, str]


class Directives(TypedDict):
	headers: dict[str, str]
	query: dict[str, str]
	socketio: SocketIODirectives


class Prerender(TypedDict):
	views: dict[str, ServerInitMessage | ServerNavigateToMessage | None]
	directives: Directives
